### PR TITLE
Clang Static Analyzer for CHERI

### PR DESCRIPTION
**Pull request by eupharina**
This is the port for Clang Static Analyzer for Morello. I used `llvm-morello` port as a template but had to add a few auxiliary scripts:

1. `ccc-analyzer` and `cxx-analyzer` wrappers are installed to `LOCALBASE/libexec` (similar to the wrappers installed by `llvm-base`) to build code with purecap ABI for analysis by default.

CSA works by overriding the path to the original compiler with `c**-analyzer` wrapper to capture all the compiler invocations and their arguments. When building a project for analysis using `cc` wrapper to add purecap flags to the compiler, CSA doesn't capture these flags, as they don't appear in the original compiler's arguments list but are added later by `cc`.
Therefore, CSA builds code for analysis with `-target-abi aapcs` even though the original compiler invocation eventually builds project code for purecap.

```
LOCALBASE/llvm-csa/libexec/ccc-analyzer <args> 	
	      |	-> cc <args> 
	      |		-> LOCALBASE/bin/clang -march=morello -mabi=purecap <args> 	
	      |			-> LOCALBASE/llvm-morello/bin/clang -cc1 -target-abi purecap <cc1_args>
	      |	-> LOCALBASE/llvm-csa/bin/clang <args> ->
				-> LOCALBASE/llvm-csa/bin/clang -cc1 -target-abi aapcs <cc1_args>
```

`c**-analyzer` wrapper adds purecap flags to `c**-analyzer` argument list, which then get propagated to the original compiler (`cc`) and analyser's clang, making the latter build and analyse code for purecap ABI.

```
LOCALBASE/libexec/ccc-analyzer <args> -> 
	LOCALBASE/llvm-csa/libexec/ccc-analyzer -march=morello -mabi=purecap <args> 
	      |	-> cc -march=morello -mabi=purecap <args> 
	      |		-> LOCALBASE/bin/clang -march=morello -mabi=purecap -march=morello -mabi=purecap <args>
	      |			-> LOCALBASE/llvm-morello/bin/clang -cc1 -target-abi purecap  <cc1_args>
	      |	-> LOCALBASE/llvm-csa/bin/clang -march=morello -mabi=purecap <args>
				-> LOCALBASE/llvm-csa/bin/clang -cc1 -target-abi purecap <cc1_args>
```
			
2. `scan-build-csa` script is installed to `LOCALBASE/bin`. This is a copy of `LOCALBASE/llvm-csa/bin/scan-build` patched after installation to use `LOCALBASE/llvm-csa/bin` to look for the analyser's compiler and scripts (except for `c**-analyzer`s, which are located by relative paths leading to wrappers mentioned above). See `scan-build-base.patch.in`.

I am unsure about a few things regarding the port:

1. I added to the `pkg-plist` only the files that CSA needs, making `llvm-csa` package size to be ~30MB vs. ~130MB for `llvm-morello`. Is it OK if `STAGEDIR` contains files that are not included in `pkg-plist`, or must I clean it somehow?

2. What would be a proper revision of Morello LLVM to rebase all my CSA patches onto (probably the one that will be used in the upcoming release)?

Thanks!

### CODE REVIEW

File: devel/Makefile [hunk 1]
```diff
@@ -1434,6 +1434,7 @@
     SUBDIR += llvm
     SUBDIR += llvm-base
     SUBDIR += llvm-cheri
+    SUBDIR += llvm-csa
```
Comment (kwitaszczyk) [inline devel/Makefile:1437]: What about `llvm-morello-csa`? This would be consistent with `llvm-morello` as the csa port extends it, and would match https://github.com/rems-project/llvm-project/tree/morello-csa.

File: devel/llvm-cheri/Makefile [hunk 2]
```diff
@@ -46,7 +46,7 @@ SHEBANG_FILES=	\
 		llvm/utils/lit/lit.py \
 		llvm/utils/llvm-lit/llvm-lit.in
 
-SUB_FILES=	llvm-wrapper.sh
+SUB_FILES+=	llvm-wrapper.sh
```
Comment (kwitaszczyk) [inline devel/llvm-cheri/Makefile:49]: I'd extend the commit message to explain why it's needed, e.g.:
> Append to `SUB_FILES` rather than overwrite it to allow ports including `devel/llvm-cheri/Makefile` to extend this list.

File: devel/llvm-morello-csa/Makefile.MAN1SRCS
Comment (kwitaszczyk) [inline devel/llvm-morello-csa/Makefile.MAN1SRCS:1]: Is this file needed?

**Comment by kwitaszczyk** [comment]
@eupharina Could you convert this to a regular PR?
@brooksdavis Could you have a look at this PR? The change in devel/llvm-cheri would require your approval. Would it have to be upstreamed to FreeBSD first?

File: devel/llvm-morello-csa/files/csa-wrapper.sh.in [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+	arch_cflags="-march=morello -mabi=purecap $tls_flags $vararg_flags"
+	arch_objdump_flags="--mattr=+morello"
+	;;
+riscv64)
+	arch_cflags="-march=rv64gcxcheri -mabi=lp64d -mno-relax"
+	;;
+riscv64c)
+	arch_cflags="-march=rv64gcxcheri -mabi=l64pc128d -mno-relax"
+	;;
```
[the remark this answers is not in the document]
Comment (eupharina) [reply devel/llvm-morello-csa/files/csa-wrapper.sh.in]: Removed these lines.

File: devel/llvm-morello-csa/files/csa-wrapper.sh.in [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	arch_objdump_flags="--mattr=+morello"
+	;;
+aarch64c)
+	if [ "$CHERIBSD_VERSION" -le 20220314 ]; then
+		tls_flags="-femulated-tls"
+		vararg_flags=
+	else
+		tls_flags=
+		vararg_flags="-Xclang -morello-vararg=new"
+	fi
```
[the remark this answers is not in the document]
Comment (eupharina) [reply devel/llvm-morello-csa/files/csa-wrapper.sh.in]: Changed to match wrapper in `llvm-base`. I made a separate fixup commit, happy to squash it once it looks fine.

**Comment by kwitaszczyk** [comment]
@eupharina Brooks has added the change from devel/llvm-cheri/Makefile in upstream. Could you rebase this PR onto latest `main`?

**Comment by eupharina** [comment]
I added [one more change to devel/llvm-cheri](https://github.com/CTSRD-CHERI/cheribsd-ports/pull/109/commits/5ed36818139d8b17066be65660f1860571dc04b4), I guess this should not be squashed. Otherwise, I think everything else should be squashed into a single commit.

**Comment by kwitaszczyk** [comment]
> I added [one more change to devel/llvm-cheri](https://github.com/CTSRD-CHERI/cheribsd-ports/pull/109/commits/5ed36818139d8b17066be65660f1860571dc04b4), I guess this should not be squashed. Otherwise, I think everything else should be squashed into a single commit.

I agree.

**Comment by brooksdavis** [comment]
> @brooksdavis Do you want to upstream [5ed3681](https://github.com/CTSRD-CHERI/cheribsd-ports/commit/5ed36818139d8b17066be65660f1860571dc04b4) first?

Sure, I can do that.

**Comment by brooksdavis** [comment]
> > @brooksdavis Do you want to upstream [5ed3681](https://github.com/CTSRD-CHERI/cheribsd-ports/commit/5ed36818139d8b17066be65660f1860571dc04b4) first?
> 
> Sure, I can do that.

I've done this and pushed a cherry-pick of it.

**Comment by kwitaszczyk** [comment]
@brooksdavis Thank you!

@eupharina Could you rebase this PR and squash appropriate commits? Once it's done, I can merge this PR.

File: devel/llvm-morello-csa/Makefile [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+GH_TAGNAME=	${LLVM_COMMIT}
+
+MASTERDIR=	${.CURDIR}/../llvm-cheri
+
+OPTIONS_EXCLUDE=	LLD LIT STATIC_LIBS DOCS
+OPTIONS_SLAVE=		CLANG CSA
+
+LLVM_TARGETS=	AArch64;ARM;BPF;X86
+
+.if exists(${.CURDIR}/files)
```
Comment (jrtc27) [inline devel/llvm-morello-csa/Makefile:26]: Why do you need the condition?

Comment (eupharina) [reply devel/llvm-morello-csa/Makefile]: I guess I don't, so I just removed it

File: devel/llvm-morello-csa/files/csa-wrapper.sh.in [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+ccc-analyzer|c++-analyzer)
+	;;
+*)
+	err invalid tool $tool
+	;;
+esac
+
+realtool=${LLVM_PREFIX}/libexec/${tool}
+
+# CheriBSD: assume that if we're on a CHERI architecture we want either
```
[the remark this answers is not in the document]
Comment (eupharina) [reply devel/llvm-morello-csa/files/csa-wrapper.sh.in:82]: I agree. However, I can't think of a simple and neat way to reuse the code from the original `wrapper.sh` without changing llvm-base port. I can at lest make the check against the latest `$CHERIBSD_VERSION` explicit, so it will crash/warn when new unsupported version arrives. Anyway, I'm open to suggestions.

File: devel/llvm-morello-csa/Makefile (hunk 5 above)
Comment (brooksdavis) [reply devel/llvm-morello-csa/Makefile]: In #142 I've subsumed this include into devel/llvm-cheri (where a different conditional is needed) so once that is merged this can be dropped entirely.

File: devel/llvm-morello-csa/files/xtoolchain.mk.in [hunk 7]
```diff
@@ -0,0 +1,6 @@
+XCC=%%PREFIX%%/llvm%%LLVM_SUFFIX%%/bin/clang
```
[the remark this answers is not in the document]
Comment (brooksdavis) [reply devel/llvm-morello-csa/files/xtoolchain.mk.in:1]: I've now added an XTOOLCHAIN option in #142 that you can add to OPTIONS_EXCLUDE

File: devel/llvm-morello-csa/Makefile (hunk 5 above)
Comment (eupharina) [reply devel/llvm-morello-csa/Makefile]: Just want to note that with changes from #142 there is no way for a derived port to opt-out from `../llvm-cheri/files/llvm*` patches. I guess with this port it's fine.

Comment (brooksdavis) [reply devel/llvm-morello-csa/Makefile]: I'll try to keep them to things that should be universal.

File: devel/llvm-morello-csa/Makefile [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+MAINTAINER=	irina.dudina@ed.ac.uk
+COMMENT=	Clang Static Analyzer for the Arm Morello architecture
+WWW=		https://github.com/rems-project/llvm-project
+
+DISTINFO_FILE=	${.CURDIR}/distinfo
+DESCR=		${.CURDIR}/pkg-descr
+PLIST=		${.CURDIR}/pkg-plist
+FILESDIR=	${.CURDIR}/files
+
+USE_GITHUB=	yes
```
Comment (brooksdavis) [inline devel/llvm-morello-csa/Makefile:14]:
```suggestion
```

File: devel/llvm-morello-csa/Makefile [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+GH_PROJECT=	llvm-project
+GH_ACCOUNT=	rems-project
+GH_TAGNAME=	${LLVM_COMMIT}
```
Comment (brooksdavis) [inline devel/llvm-morello-csa/Makefile:17]:
```suggestion
```

File: devel/llvm-morello-csa/Makefile [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+OPTIONS_EXCLUDE=	LLD LIT STATIC_LIBS DOCS
+OPTIONS_SLAVE=		CLANG CSA
+
+LLVM_TARGETS=	AArch64;ARM;BPF;X86
+
+EXTRA_PATCHES=	${.CURDIR}/files
+
+.include "${.CURDIR}/Makefile.snapshot"
+
```
Comment (brooksdavis) [inline devel/llvm-morello-csa/Makefile:29]:
```suggestion

```

File: devel/llvm-morello-csa/Makefile [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+SUB_FILES=	csa-wrapper.sh
+
+post-patch-CSA-on:
+	@${REINPLACE_CMD} -e 's|%%LLVM_PREFIX%%|${LLVM_PREFIX}|g' \
+		${WRKSRC}/clang/tools/scan-build/bin/scan-build
+
+CSA_COMMANDS=	ccc-analyzer c++-analyzer
```
Comment (brooksdavis) [inline devel/llvm-morello-csa/Makefile:36]: I guess this can't go in Makefile.COMMANDS because it's needed in the `.for` loop?

**Review by brooksdavis — COMMENTED** [review]
I've made a few more changes to #142 which fix GH_ issues and slightly reduce duplication.  Sorry for the churn.

File: devel/llvm-morello-csa/Makefile (hunk 11 above)
Comment (eupharina) [reply devel/llvm-morello-csa/Makefile:30]: Yes, `CSA_COMMANDS` variable is for this loop only. I tried to keep this code similar to llvm-base/Makefile

**Comment by kwitaszczyk** [comment]
I think all of the issues except for the duplicate code in `csa-wrapper.sh.in` are resolved now. I agree with Jessica's point that it would be better to make use of llvm-base instead. However, I think it will require changes to llvm-base itself. For now, we can use the port as it is and improve it separately.
